### ISSUE TITLE
Possible to set deep "extra" config values via the command-line interface? (workarounds included)

### ISSUE THREAD

**Issue by TravisCarden** [composer/composer#8353]
tl;dr: I need to set (>3 levels) deep "extra" config values from the command-line. Is it possible to do this with Composer, or must one resort to external tools? Could the 3 levels depth limit be increased?

The documentation for [modifying extra values via the command-line interface](https://getcomposer.org/doc/03-cli.md#modifying-extra-values) states, "dots indicate array nesting, a max depth of 3 levels is allowed though". Unfortunately, I frequently need to set deeper values--for [cweagans/composer-patches](https://github.com/cweagans/composer-patches) config, for example, which looks like this:

```json
{
  "extra": {
    "patches": {
      "vendor/project": {
        "Patch title": "http://example.com/url/to/patch.patch"
      }
    }
  }
}
```

Ideally, I would create that like this:

```bash
composer config "extra.patches.vendor/project.Patch title" http://example.com/url/to/patch.patch
```

But of course that results in this instead:

```json
{
  "extra": {
    "patches": {
      "vendor/project.Patch title": "http://example.com/url/to/patch.patch"
    },
  }
}
```

Is it currently possible to accomplish this some other way via the command-line interface? If not, could it be _made_ possible--perhaps by deepening the maximum depth limit or making it configurable via argument (e.g., `composer config --max-depth=5`)?

## Workarounds for weary travelers

- For the Composer patches use case specifically, [using an external patch file](https://github.com/cweagans/composer-patches#using-an-external-patch-file) should probably be the cleanest solution, though I have not personally succeeded in making it work, e.g.:

    ```bash
    composer config extra.patches-file local/path/to/your/composer.patches.json
    ```

- Another good command-line solution (which is installed by default in Travis CI build environments) is [jq](https://stedolan.github.io/jq/), e.g.:

    ```bash
    jq '.extra.patches = {"vendor/project": {"Patch title": "http://example.com/url/to/patch.patch"}}' composer.json > temp.json && mv temp.json composer.json
    ```

- Within a PHP script, it can be accomplished like this:

    ```php
    $json_file = new \Composer\Json\JsonFile('composer.json');
    $data = $json_file->read();
    $data['extra']['patches']['vendor/project']['Patch title'] = 'http://example.com/url/to/patch.patch';
    $json_file->write($data);
    ```

After any of these you'll need to force Composer to update the lock file and (in the case of Composer patches, specifically) redownload and patch the affected project:

```bash
rm -rf vendor/vendor/project
composer update --lock
```

**Comment by Seldaek** [comment]
The main issue with this IIRC is that it conflicts with repositories and auths having domain names in their third level keys, so if we split by dot further it ends up breaking things. It probably is technically feasible to do this selectively depending on the key, or to allow splitting using another character than dots perhaps. The other issue is the JsonManipulator class (which is used to patch json files without modifying white-space, ordering and formatting) is not presently able to go deeper than 3 I believe.

Just pointers for someone interested in possibly fixing this.. but tbh I'd say `jq` is a better option as it's a generic json editing tool and that's not something we really want to fully reimplement in Composer.

**Comment by effulgentsia** [comment]
> The main issue with this IIRC is that it conflicts with repositories and auths having domain names in their third level keys, so if we split by dot further it ends up breaking things.

One of the suggestions in the OP is to add a command line option: e.g., --max-depth.

Another use-case for where this would be helpful is [composer/installers](https://github.com/composer/installers), because `installer-paths` requires a depth of 4. For example, the following would be a useful capability:
`composer config --max-depth=4 extra.installer-paths.your/custom/path/{\$name}/.0 vendor:my_organization`

> tbh I'd say jq is a better option as it's a generic json editing tool and that's not something we really want to fully reimplement in Composer
Certainly there's some cutoff where Composer shouldn't reimplement all of jq. However, the `composer config` command is a helpful one for simple tasks, particularly because you can write INSTALL / README instructions based on that command, rather than having to write instructions that say "edit composer.json in this way..." or that bias for a particular 3rd party json editor.

**Comment by effulgentsia** [comment]
Another thought, if it's easier/simpler, is perhaps to add a `--json` option to specify that the value should be treated as json already, and that way, structured data could be passed to it. For example,

`composer config --json extra.installer-paths.your/custom/path/{\$name}/ '["vendor:my_organization"]'`

#8195 is a similar issue unless that one gets scoped down to just different scalar types.

**Comment by karol-haltenberger** [comment]
It seems that the json string method - as mentioned for repositories - doesn't work here either.
How about allowing that for the extra section?

**Comment by Seldaek** [comment]
https://github.com/composer/installers/pull/449 is merged now which alleviates some of the problem. I think the --json alternative is also not bad, but not sure exactly how easy it would be to implement. If someone wants to have a go though feel free.

**Comment by effulgentsia** [comment]
Thanks, @Seldaek for the encouragement to proceed with that. I opened https://github.com/composer/composer/pull/8779, which I'm sure needs to more work before it can be merged, but it basically works.

**Comment by Seldaek** [comment]
Fixed by https://github.com/composer/composer/pull/8779
